common/beerocks/bwl/shared/nl80211_client_impl.cpp

```diff
@@ -549,6 +549,103 @@ bool nl80211_client_impl::get_sta_info(const std::string &interface_name,
+        [&](struct nl_msg *msg) -> bool {
+            nla_put_u32(msg, NL80211_ATTR_IFINDEX, iface_index);
+
+            return true;
+        },
+        [&](struct nl_msg *msg) {
+            struct nlattr *tb[NL80211_ATTR_MAX + 1];
+            struct genlmsghdr *gnlh = static_cast<genlmsghdr *>(nlmsg_data(nlmsg_hdr(msg)));
+            struct nlattr *sinfo[NL80211_SURVEY_INFO_MAX + 1];
+            survey_info survey_info;
+
+            // Parse the netlink message
+            if (nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0),
+                          NULL)) {
+                LOG(ERROR) << "Failed to parse netlink message!";
+                return;
+            }
+
+            if (!tb[NL80211_ATTR_SURVEY_INFO]) {
+                LOG(ERROR) << "Survey data missing!";
+                return;
+            }
+
+            // Parse nested survey info
+            if (nla_parse_nested(sinfo, NL80211_SURVEY_INFO_MAX, tb[NL80211_ATTR_SURVEY_INFO],
+                                 survey_policy)) {
+                LOG(ERROR) << "Failed to parse nested survey attributes!";
+                return;
+            }
+
+            survey_info.in_use = sinfo[NL80211_SURVEY_INFO_IN_USE];
+
+            if (sinfo[NL80211_SURVEY_INFO_FREQUENCY]) {
+                survey_info.frequency_mhz = nla_get_u32(sinfo[NL80211_SURVEY_INFO_FREQUENCY]);
+            } else {
+                LOG(DEBUG) << "NL80211_SURVEY_INFO_FREQUENCY attribute is missing";
+            }
+
+            if (sinfo[NL80211_SURVEY_INFO_NOISE]) {
+                survey_info.noise_dbm = (int8_t)nla_get_u8(sinfo[NL80211_SURVEY_INFO_NOISE]);
+            } else {
+                LOG(DEBUG) << "NL80211_SURVEY_INFO_NOISE attribute is missing";
+            }
+
+            if (sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME]) {
+                survey_info.time_on_ms = nla_get_u64(sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME]);
+            }
+
+            if (sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_BUSY]) {
+                survey_info.time_busy_ms =
+                    nla_get_u64(sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_BUSY]);
+            }
+
+            if (sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_EXT_BUSY]) {
+                survey_info.time_ext_busy_ms =
+                    nla_get_u64(sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_EXT_BUSY]);
+            }
+
+            if (sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_RX]) {
+                survey_info.time_rx_ms = nla_get_u64(sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_RX]);
+            }
+
+            if (sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_TX]) {
+                survey_info.time_tx_ms = nla_get_u64(sinfo[NL80211_SURVEY_INFO_CHANNEL_TIME_TX]);
+            }
+
+            if (sinfo[NL80211_SURVEY_INFO_TIME_SCAN]) {
+                survey_info.time_scan_ms = nla_get_u64(sinfo[NL80211_SURVEY_INFO_TIME_SCAN]);
+            }
+
+            survey_info_list.push_back(survey_info);
```

Comment:
(nitpick) I think we should take the habit of using emplace_back instead of push_back. In this case it doesn't make a difference because none of the fields have move semantics, but if there would be e.g. a string in there, it would be more efficient.

Reply:
Done